### ISSUE TITLE
Assignment

### ISSUE THREAD

**Issue by DNatanaelK** [brarcher/budget-watch#182]
Hi there!
Willing to add some features to the current application of Budget-Watch for my university assignment. 
There will be 3 features, Auto-complete upon filling up details, Notify warning message if budget is over the limit, able to purge all transactions (expenses and revenues).
Thanks!

**Comment by brarcher** [comment]
Thanks for the interest!

> Auto-complete upon filling up details

If you are interested in an auto-complete feature, take a look at https://github.com/brarcher/budget-watch/issues/110 and the discussion in https://github.com/brarcher/budget-watch/pull/152. 

> Notify warning message if budget is over the limit

Hm. Because transactions only occur when a user explicitly adds one, I'm not sure how relevant a warning would be. Do you mean a warning notification? You may need more justification on this point.

> able to purge all transactions (expenses and revenues)

The only feature which is close to this is the ability to delete receipts from a date range. If one wanted to purge all data, why not clear the app's data from the Settings menu? You may need more justification on this one as well.

If there are other features which you may be interested in working on, there are some other options in the Issues section which may be worth considering. If you would like a suggestion, I could point you to one or two.

**Comment by brarcher** [comment]
> This feature is to be added for only transactions and revenues, for instance if users have too much transactions and revenues that are old and not used anymore, with a single click it is possible to delete them.

Transaction data does not take up that much space. I'm not sure how valuable such a feature would be, especially given that someone can wipe the data in the app fairly easily.

> Just a pop-up warning notification, if a particular budget has been over the limit.

Might put this one on a back burner. I do not have a lot of confidence that it will be very useful to users. Or, maybe that is just my specific use case talking. (:

> Maybe there are other suggestions? ... are there any bugs within the application?

The only bug I'm aware of is https://github.com/brarcher/budget-watch/issues/156. I would say the nicest feature which could be contributed besides auto complete is https://github.com/brarcher/budget-watch/issues/61, which involves changing the database and migrating data.

**Comment by DNatanaelK** [comment]
Okay! noted. 

> The only bug I'm aware of is #156.

Has this bug been fixed yet? 

> I would say the nicest feature which could be contributed besides auto complete is #61, which involves changing the database and migrating data.

Isn't this used to fixed the Budget Name? 

I have also seen the issue #152, it seems like it is only for the transaction activity?

**Comment by brarcher** [comment]
> Has this bug been fixed yet?

Nope.

> Isn't this used to fix the Budget Name bug?

Correct. That would need to be fixed in order to fix the Budget Name bug.

> I have also seen the issue #152, it seems like it is only for the transaction activity?

#152 is the Auto-complete feature request. As Budgets are added so infrequently, it would be more valuable to have the auto-complete on Transactions rather than Budgets.
 
> and there are 2 set of codes one is generalized and one is the name field, how are they different?

If you can give me function names or line numbers I can speak to them. If you see code which is fairly similar in a few places it may be an opportunity to refactor. (:

**Comment by DNatanaelK** [comment]
> Correct. That would need to be fixed in order to fix the Budget Name bug.

I see, does the issue #61 solves the Budget Name bug?

> #152 is the Auto-complete feature request. As Budgets are added so infrequently, it would be more valuable to have the auto-complete on Transactions rather than Budgets.

I agree on this too!

> If you can give me function names or line numbers I can speak to them. If you see code which is fairly similar in a few places it may be an opportunity to refactor. (:

What I meant was, I think TransactionViewActivity.java was changed to "Add autocomplete to Name field" and then overwritten again to "Generalize autocomplete fields in transaction activity" if I'm not mistaken?

Thanks!

**Comment by brarcher** [comment]
> I see, does the issue #61 solves the Budget Name bug?

There is probably something else broke in the Budget activity, as the budget renaming did work at one point (though when you rename the transactions would not follow the new name). However, without the database normalization, renaming a budget would not do what you expect.

> What I meant was...

Oh, you are talking about the commits in #152. Note that the solution does not need to include those changes. You are welcome to build off of change from @clinchergt, or to propose your own solution. Please keep in mind the discussion in #152 as to what a solution should do. You will also need unit tests to test the new functionality.

**Comment by brarcher** [comment]
> are there perhaps any bugs other than the budget name?

I looked through the issues and there are none that I know of. The other pending issues are feature requests or enhancements.

**Comment by DNatanaelK** [comment]
> I looked through the issues and there are none that I know of. The other pending issues are feature requests or enhancements.

Noted.

> Oh, you are talking about the commits in #152. 

I tried implementing this Issue to the app and for the first one which is "Add autocomplete to Name field" it worked but when I tried implementing the second one which is "Generalize autocomplete fields  in transaction activity" it did not work, even the names are not able to auto-complete. Was this issue experienced too? Or did I make some mistakes of my own.

Thanks!

**Comment by brarcher** [comment]
I've not tried the changes, as they would need to be implemented as storing data in the database instead.

**Comment by DNatanaelK** [comment]
Noted. 

Sorry for the late response, and thanks for your concern and replies! 

Is it okay if I comment here if I have some questions?

**Comment by brarcher** [comment]
It may be better if you comment here for general questions as well as about what you may be interested in tackling. If you would like to work on a specific issue comment either in the related issue or a pull request that you propose.

**Comment by brarcher** [comment]
The [DBHelper](https://github.com/brarcher/budget-watch/blob/master/app/src/main/java/protect/budgetwatch/DBHelper.java) class has methods used to manipulate the database. There currently is not a method for deleting all transactions. If you are interested in creating a new method to delete all transactions, I suggest reading the documentation on the [DELETE](https://www.tutorialspoint.com/sqlite/sqlite_delete_query.htm) command here, and practicing on an emulator with the sqlite3 tool to get the command just right. 

If you are looking into adding the feature to delete all transactions, I'm not convinced that the feature would be that useful as is. I could see a case where removing all transactions before a certain date could be useful. There is a feature similar to this, something that will [delete all receipts](https://github.com/brarcher/budget-watch/blob/master/app/src/main/java/protect/budgetwatch/TransactionActivity.java#L161) (images) before a certain date. You could model the selective transaction purge after that.

**Comment by DNatanaelK** [comment]
> If you are interested in creating a new method to delete all transactions, I suggest reading the documentation on the DELETE command here, and practicing on an emulator with the sqlite3 tool to get the command just right.

Okay I will look into this! I tried working around the database, didn't seem to work yet, will try it again!

> There is a feature similar to this, something that will delete all receipts (images) before a certain date. You could model the selective transaction purge after that. 

Yeap, I have seen this! I tried working around this too, but couldn't still make the Delete all Transactions to work yet. 

> If you are looking into adding the feature to delete all transactions, I'm not convinced that the feature would be that useful as is.

Are there maybe any other features that is you can suggest? It can be anything for the application. And for the bug fix, it has to do with the database too? I tried working around #61 but I could not catch the important points of it and the Bug was not fixed. 

> I could see a case where removing all transactions before a certain date could be useful. 

Are there any developments regarding to this? Or perhaps anyone interested to develop this?

Thanks for the reply and time!

**Comment by brarcher** [comment]
> For the bug fix, it has to do with the database too? I tried working around #61 but I could not catch the important points of it and the Bug was not fixed.

I think that the bug where the budget name cannot be changed is something unrelated to the database normalization. The budget should be able to be renamed. However, without fixing the database normalization, renaming a budget will effectively be like creating a new budget, in that it will not have any transactions follow it.

So, let me ask about this university assignment. Is the purpose to contribute to an open source project in some capacity? How much time are you expected to dedicate? If you are looking for a bug which is probably less involved, I have a few other apps you could contribute to. Here are some examples of bugs: https://github.com/brarcher/loyalty-card-locker/issues/233, https://github.com/brarcher/loyalty-card-locker/issues/167. Additionally, https://github.com/brarcher/budget-watch/issues/183 was logged recently and you could investigate that if you want.

**Comment by DNatanaelK** [comment]
> I think that the bug where the budget name cannot be changed is something unrelated to the database normalization. The budget should be able to be renamed. However, without fixing the database normalization, renaming a budget will effectively be like creating a new budget, in that it will not have any transactions follow it.

Yeah, I tried working around the database and it did not solve the bug.

> So, let me ask about this university assignment. Is the purpose to contribute to an open source project in some capacity? How much time are you expected to dedicate?

I am supposed to add features into an application and solve at least one known bug, for the adding of features part I am almost done, but for the bug fixing I am not quite there yet. It is to be done by next week. 

>  If you are looking for a bug which is probably less involved, I have a few other apps you could contribute to. Here are some examples of bugs: brarcher/loyalty-card-locker#233, brarcher/loyalty-card-locker#167. 

Thanks for the suggestion, but I think I have to stick to one application, which is the Budget Watch.

> Additionally, #183 was logged recently and you could investigate that if you want.

Yes I have seen this, but I have not reproduce the bug yet. Is this bug is about layout styles? for the intro pages? I think the user that issued this has too less of information of the bug itself. 

> brarcher/loyalty-card-locker#233, brarcher/loyalty-card-locker#167

I just realized that you have other applications developed! I will check and try them out too!

Thanks for the time and reply!

**Comment by DamageIncorporated** [comment]
> > Notify warning message if budget is over the limit
> 
> Hm. Because transactions only occur when a user explicitly adds one, I'm not sure how relevant a warning would be. Do you mean a warning notification? You may need more justification on this point.

I would consider a warning dialog useful too. when adding a transaction I may be not aware that this is the one too exceed the budget it belongs to. 
I agree that I won't cancel the transaction at this point: The money is already spent but it would be good to be notified that I've reached or passed my limit for this month and it's time to hit the spending-break...
I've added #185 today to tackle the problem from another angle.
